app/code/Magento/Sales/Model/Order.php

```diff
@@ -614,30 +614,49 @@ public function canCreditmemo()
         if ($this->hasForcedCanCreditmemo()) {
             return $this->getForcedCanCreditmemo();
         }
-
-        if ($this->canUnhold() || $this->isPaymentReview()) {
-            return false;
-        }
-
-        if ($this->isCanceled() || $this->getState() === self::STATE_CLOSED) {
+        $isOnHoldReview = $this->canUnhold() || $this->isPaymentReview();
+        $isOrderCanceled = $this->isCanceled() || $this->getState() === self::STATE_CLOSED;
+        if ($isOnHoldReview || $isOrderCanceled) {
             return false;
         }
-
         /**
          * We can have problem with float in php (on some server $a=762.73;$b=762.73; $a-$b!=0)
          * for this we have additional diapason for 0
          * TotalPaid - contains amount, that were not rounded.
          */
         $totalRefunded = $this->priceCurrency->round($this->getTotalPaid()) - $this->getTotalRefunded();
-        if (abs($totalRefunded) < .0001) {
-            return false;
+        if (abs($this->getGrandTotal()) < .0001) {
+            return $this->canCreditmemoforZeroTotal($totalRefunded);
         }
+        $isRefundZero = abs($totalRefunded) < .0001;
         // Case when Adjustment Fee (adjustment_negative) has been used for first creditmemo
-        if (abs($totalRefunded - $this->getAdjustmentNegative()) < .0001) {
+        $hasAdjustmentFee = abs($totalRefunded - $this->getAdjustmentNegative()) < .0001;
+        $hasActinFlag = $this->getActionFlag(self::ACTION_FLAG_EDIT) === false;
+        if ($isRefundZero || $hasAdjustmentFee || $hasActinFlag) {
             return false;
         }
+        return true;
+    }
 
-        if ($this->getActionFlag(self::ACTION_FLAG_EDIT) === false) {
+    /**
+     * @param $totalRefunded
+     * @return bool
+     *
+     */
+    public function canCreditmemoforZeroTotal($totalRefunded)
+    {
+        $totalPaid = $this->getTotalPaid();
+        //check if total paid is less than grandtotal
+        $checkAmtTotalPaid = $totalPaid <= $this->getGrandTotal();
+        //case when amount is due for invoice
+        $dueAmountCondition = $this->canInvoice() && ($checkAmtTotalPaid);
+        //case when paid amount is refunded and order has creditmemo created
+        $paidAmtIsRefunded = $this->getTotalRefunded() == $totalPaid && count($this->getCreditmemosCollection()) > 0;
+        $state1 = $dueAmountCondition || $paidAmtIsRefunded;
```

Comment:
Please use meaningful variable names, avoid generic names like `$state1` and `$state2`